### PR TITLE
INTMDB-28: Added Event Triggers Realm

### PR DESCRIPTION
**Pull request by coderGo93**
## Description

- Added resource, datasources and its tests for Event Triggers Realm
- Modified config provider to get client realm or atlas
- Modified all resources/datasources for atlas with one line which is the way to get atlas client, this is why appears many files changed

Link to any related issue(s):

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the Terraform contribution guidelines
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments
Waiting for a [PR](https://github.com/10gen/go-client-mongodb-realm/pull/3) of realm client to be merged then pull the changes so it can pass the normal tests

### CODE REVIEW

**Comment by themantissa** [comment]
@coderGo93 added apix to review. Reminder to share PRs in Slack as we discussed.  Also tests failed.

**Comment by coderGo93** [comment]
@themantissa thank you and yes because I'm a waiting a pr in realm client to be merged to master and then pull the changes in go modules and it will pass the test

**Comment by themantissa** [comment]
@tsedgwick though you'd want to see (and perhaps comment on/review) the PR for the work in Terraform.

File: go.sum [hunk 1]
```diff
@@ -765,10 +765,15 @@ go.mongodb.org/atlas v0.7.2 h1:wB3+hP71t3mK+JOSrjBFbrzb5MsZRzDtZlpEKp58KK0=
...
 go.mongodb.org/atlas v0.7.3-0.20210302081736-7ccf7e1f4351/go.mod h1:CIaBeO8GLHhtYLw7xSSXsw7N90Z4MFY87Oy9qcPyuEs=
+go.mongodb.org/atlas v0.8.0/go.mod h1:MMWDsc2akjTDSG4tVQrxv/82p3QbBnqeELbtTl45sbg=
 go.mongodb.org/atlas v0.8.1-0.20210528214129-330aa9199dbf h1:B9IN5pGOtgtgmx8jjtuP9Gx6PPFWIIKx3AKQKYKcxJM=
 go.mongodb.org/atlas v0.8.1-0.20210528214129-330aa9199dbf/go.mod h1:MMWDsc2akjTDSG4tVQrxv/82p3QbBnqeELbtTl45sbg=
 go.mongodb.org/atlas v0.8.1-0.20210604212331-3f489d7a4eed h1:qPaT/RvfXXKEJ2fTNckgPKoqfg4oLgJFj7exXUM8mUQ=
 go.mongodb.org/atlas v0.8.1-0.20210604212331-3f489d7a4eed/go.mod h1:MMWDsc2akjTDSG4tVQrxv/82p3QbBnqeELbtTl45sbg=
+go.mongodb.org/realm v0.0.0-20210604190232-4d3dfc67ed71 h1:2jY3qOwzS8AQMRtuGxs8y/k1FRGWQHfvrrtEk0X+61Y=
+go.mongodb.org/realm v0.0.0-20210604190232-4d3dfc67ed71/go.mod h1:Tru1+aHka6+4uXYigPKoNySbNWgtkfsanVddwmcqvOM=
+go.mongodb.org/realm v0.0.0-20210618220639-e70c919266f2 h1:3ukcBKIOun3QRMZzdVRCU/7uVhjR7XQsdD/cCH5V8So=
+go.mongodb.org/realm v0.0.0-20210618220639-e70c919266f2/go.mod h1:Tru1+aHka6+4uXYigPKoNySbNWgtkfsanVddwmcqvOM=
```
Comment (threebee) [inline go.sum]: [q] Why does go.sum mentions 2 versions of `go.mongodb.org/realm`? When only `go.mongodb.org/realm v0.0.0-20210618220639-e70c919266f2` is present in go.mod?

**Review by threebee — COMMENTED** [review]
Looking good, I have added a question for my own learning.

File: go.sum (hunk 1 above)
Comment (colm-quinn) [reply go.sum]: That looks like `go mod tidy` might need to be run to remove the duplicates, unless the reference is coming in from another lib.

Comment (coderGo93) [reply go.sum]: good eye @threebee , just pushed with vendor updated, thank you

File: mongodbatlas/data_source_mongodbatlas_event_triggers_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		appID        = os.Getenv("MONGODB_REALM_APP_ID")
+		eventResp    = realm.EventTrigger{}
+	)
+	event := realm.EventTriggerRequest{
+		Name:       acctest.RandomWithPrefix("test-acc"),
+		Type:       "DATABASE",
+		FunctionID: os.Getenv("MONGODB_REALM_FUNCTION_ID"),
+		Disabled:   pointy.Bool(false),
+		Config: &realm.EventTriggerConfig{
+			OperationTypes: []string{"INSERT", "UPDATE"},
```
Comment (themantissa) [inline mongodbatlas/data_source_mongodbatlas_event_triggers_test.go:28]: Same as above, maybe a test for a login type and a test for a database type

File: mongodbatlas/resource_mongodbatlas_event_trigger.go [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+				Required: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringInSlice([]string{"DATABASE", "AUTHENTICATION", "SCHEDULED"}, false),
```
Comment (themantissa) [inline mongodbatlas/resource_mongodbatlas_event_trigger.go:52]: Ah you are right, SCHEDULED is not in the doc is it?  Is this what you mean when you said the docs were wrong?

File: mongodbatlas/data_source_mongodbatlas_event_trigger_test.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		eventResp    = realm.EventTrigger{}
+	)
+	event := realm.EventTriggerRequest{
+		Name:       acctest.RandomWithPrefix("test-acc"),
+		Type:       "DATABASE",
+		FunctionID: os.Getenv("MONGODB_REALM_FUNCTION_ID"),
+		Disabled:   pointy.Bool(false),
+		Config: &realm.EventTriggerConfig{
+			OperationTypes: []string{"INSERT", "UPDATE"},
+			OperationType:  "LOGIN",
```
Comment (themantissa) [inline mongodbatlas/data_source_mongodbatlas_event_trigger_test.go]: OperationType Login is not needed for a DB trigger, please ensure the config includes what is required (unless you are trying to test if someone includes a parameter not needed as well as needed?)

File: mongodbatlas/resource_mongodbatlas_event_trigger_test.go [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+		eventResp    = realm.EventTrigger{}
+	)
+	event := realm.EventTriggerRequest{
+		Name:       acctest.RandomWithPrefix("test-acc"),
+		Type:       "DATABASE",
+		FunctionID: os.Getenv("MONGODB_REALM_FUNCTION_ID"),
+		Disabled:   pointy.Bool(false),
+		Config: &realm.EventTriggerConfig{
+			OperationTypes: []string{"INSERT", "UPDATE"},
+			OperationType:  "LOGIN",
```
Comment (themantissa) [inline mongodbatlas/resource_mongodbatlas_event_trigger_test.go]: Same as above, OperationType is not needed by a Database trigger.

File: website/docs/d/event_trigger.html.markdown [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+`mongodbatlas_event_trigger` describe an Event Trigger. 
+
+## Example Usage
+
+```hcl
+resource "mongodbatlas_event_trigger" "test" {
+  project_id = "PROJECT ID"
+  app_id = "APPLICATION ID"
+  name = "NAME OF THE TRINGGER"
```
Comment (themantissa) [inline website/docs/d/event_trigger.html.markdown]: Spelling error:
 "NAME OF THE TRIGGER"

File: website/docs/d/event_trigger.html.markdown [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  type = "DATABASE"
+  function_id = "1"
+  disabled = false
+  config_operation_types = ["INSERT", "UPDATE"]
+  config_operation_type = "LOGIN"
```
Comment (themantissa) [inline website/docs/d/event_trigger.html.markdown]: Two different types of triggers mixed here too, this is for an authentication trigger

File: website/docs/d/event_triggers.html.markdown [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+description: |-
+    Describes an Event Triggers.
+---
+
+# mongodbatlas_event_triggers
+
+`mongodbatlas_event_triggers` describe all Event Triggers.
+
+## Example Usage
+
```
Comment (themantissa) [inline website/docs/d/event_triggers.html.markdown:14]: any reason we don't have the example create before this like the other data source example?

File: website/docs/r/event_trigger.html.markdown [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    Provides a Event Trigger resource.
+---
+
+# mongodbatlas_event_trigger
+
+`mongodbatlas_event_trigger` provides a Event Trigger resource. 
+
+## Example Usages
+
+### Example Usage: DATABASE with function
```
Comment (themantissa) [inline website/docs/r/event_trigger.html.markdown]: I'd use the whole name in the title:
 ### Example Usage: Database Trigger with Function

File: website/docs/r/event_trigger.html.markdown [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  event_processors {
+    aws_eventbridge {
+      config_account_id = "AWS ACCOUNT ID"
+      config_region = "AWS REGIOn"
+    }
+  }
+}
+```
+
+### Example Usage: DATABASE with eventbridge
```
Comment (themantissa) [inline website/docs/r/event_trigger.html.markdown]: Same, just make it a bit easier to read:
Example Usage: Database Trigger with EventBridge

File: website/docs/r/event_trigger.html.markdown [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  event_processors {
+    aws_eventbridge {
+      config_account_id = "AWS ACCOUNT ID"
+      config_region = "AWS REGIOn"
+    }
+  }
+}
+```
+
+### Example Usage: AUTHENTICATION
```
Comment (themantissa) [inline website/docs/r/event_trigger.html.markdown]: Just add trigger: 
### Example Usage: Authentication Trigger

File: website/docs/r/event_trigger.html.markdown [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  name = "NAME OF THE TRIGGER"
+  type = "AUTHENTICATION"
+  function_id = "1"
+  disabled = false
+  config_operation_type = "LOGIN"
+  config_providers = ["anon-user"]
+}
+```
+
+### Example Usage: SCHEDULED with eventbridge
```
Comment (themantissa) [inline website/docs/r/event_trigger.html.markdown]: Same
 ### Example Usage: Scheduled with EventBridge

File: website/docs/r/event_trigger.html.markdown [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+## Argument Reference
+
+* `project_id` - (Required) The unique ID for the project to create the trigger.
+* `app_id` - (Required) The ObjectID of your application.
+* `name` - (Required) The name of the trigger.
+* `type` - (Required) The type of the trigger. Possible Values: `DATABASE`, `AUTHENTICATION`
+* `function_id` - (Optional) The ID of the function associated with the trigger.
+* `disabled` - (Optional) Default: `false` If `true`, the trigger is disabled.
+* `config_operation_types` - (Optional) The [database event operation types](https://docs.mongodb.com/realm/triggers/database-triggers/#std-label-database-events) to listen for. This must contain at least one value. Required for `DATABASE` type. Possible Values: `INSERT`, `UPDATE`, `REPLACE`, `DELETE`
```
Comment (themantissa) [inline website/docs/r/event_trigger.html.markdown]: I would move required up a bit:
`config_operation_types` - (Optional) Required for `DATABASE` type.  The [database event operation types](https://docs.mongodb.com/realm/triggers/database-triggers/#std-label-database-events) to listen for. This must contain at least one value. Possible Values: `INSERT`, `UPDATE`, `REPLACE`, `DELETE`

File: website/docs/r/event_trigger.html.markdown [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+* `config_operation_type` - (Optional) The [authentication operation type](https://docs.mongodb.com/realm/triggers/authentication-triggers/#std-label-authentication-event-operation-types) to listen for. Required for `AUTHENTICATION` type. Possible Values: `LOGIN`, `CREATE`, `DELETE`
```
Comment (themantissa) [inline website/docs/r/event_trigger.html.markdown]: Same on the required, move up (though it is weird to have after the Terraform option/required note but I think people get that hopefully.  Can you do the same for all here with a requirement? Example:

* `config_operation_type` - (Optional) Required for `AUTHENTICATION` type.  The [authentication operation type](https://docs.mongodb.com/realm/triggers/authentication-triggers/#std-label-authentication-event-operation-types) to listen for. Possible Values: `LOGIN`, `CREATE`, `DELETE`

File: website/docs/r/event_trigger.html.markdown [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+* `config_schedule` - (Optional) A [cron expression](https://docs.mongodb.com/realm/triggers/cron-expressions/) that defines the trigger schedule.
+* `event_processors` - (Optional) An object where each field name is an event processor ID and each value is an object that configures its corresponding event processor. The following event processors are supported: `AWS_EVENTBRIDGE` For an example configuration object, see [Send Trigger Events to AWS EventBridge](https://docs.mongodb.com/realm/triggers/eventbridge/#std-label-event_processor_example).
+* `event_processors.0.aws_eventbridge.config_account_id` - (Optional) AWS Account ID.
+* `event_processors.0.aws_eventbridge.config_region` - (Optional) Region of AWS Account.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The Terraform's unique identifier used internally for state management.
```
Comment (themantissa) [inline website/docs/r/event_trigger.html.markdown]: slight edit:
 * `id` - Terraform's unique identifier used internally for state management.

File: website/docs/r/event_trigger.html.markdown [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+  config_schedule = "*"
+}
+```
+
+## Argument Reference
+
+* `project_id` - (Required) The unique ID for the project to create the trigger.
+* `app_id` - (Required) The ObjectID of your application.
+* `name` - (Required) The name of the trigger.
+* `type` - (Required) The type of the trigger. Possible Values: `DATABASE`, `AUTHENTICATION`
```
Comment (themantissa) [inline website/docs/r/event_trigger.html.markdown]: Missing SCHEDULED

**Review by themantissa — CHANGES_REQUESTED** [review]
Overall lgtm but some small docs edits and test stuff.

File: mongodbatlas/data_source_mongodbatlas_event_triggers_test.go (hunk 2 above)
Comment (coderGo93) [reply mongodbatlas/data_source_mongodbatlas_event_triggers_test.go:28]: already has in the resource tests including with eventbridge for database, auth and schedule

File: mongodbatlas/resource_mongodbatlas_event_trigger.go (hunk 3 above)
Comment (coderGo93) [reply mongodbatlas/resource_mongodbatlas_event_trigger.go:52]: Yes and the response, the response was different than the docs, the response returns almost the same as the request with additional field like `function_name` and id

File: mongodbatlas/data_source_mongodbatlas_event_trigger_test.go (hunk 4 above)
Comment (coderGo93) [reply mongodbatlas/data_source_mongodbatlas_event_trigger_test.go]: you're right, fixed it, thank you melissa

File: website/docs/d/event_triggers.html.markdown (hunk 8 above)
Comment (coderGo93) [reply website/docs/d/event_triggers.html.markdown:14]: no reason, I added the example of resource before using data source

File: mongodbatlas/data_source_mongodbatlas_event_triggers_test.go (hunk 2 above)
Comment (themantissa) [reply mongodbatlas/data_source_mongodbatlas_event_triggers_test.go:28]: Ah, yup - thanks!

File: mongodbatlas/resource_mongodbatlas_event_trigger.go (hunk 3 above)
Comment (themantissa) [reply mongodbatlas/resource_mongodbatlas_event_trigger.go:52]: Good to know, thanks!

**Review by themantissa — APPROVED** [review]
LGTM! Thank you for the edits.

**Review by threebee — APPROVED** [review]
LGTM if test passes. I personally found this PR kind of too big to review.

**Comment by themantissa** [comment]
@threebee do you think it would have helped to do the changes for the URI in one PR and the actual Trigger add in another?   I agree the combined made for a large PR!  fyi @coderGo93

**Comment by coderGo93** [comment]
@threebee The tests for trigger are coded to be skipped because it requires to be available always for app realm since there's no a resource to create a realm app, also to load a sample dataset in cluster since it requires to link the database name and collection name, what do you think? @themantissa .

Yeah the PR is kinda big because of implementation of the second client instead of only atlas but I did that way because it seems safer to avoid any breaking since once it's merged, the other "old" PRs must be rebased to avoid an error that client type is wrong so in that way, if most of the tests of atlas passed it means it works the new implementation, obviously there's a chance that it will fail some testacc like before but it's expected since it hasn't fixed yet

**Comment by threebee** [comment]
> @threebee The tests for trigger are coded to be skipped because it requires to be available always for app realm since there's no a resource to create a realm app, also to load a sample dataset in cluster since it requires to link the database name and collection name, what do you think? @themantissa .
> 
> Yeah the PR is kinda big because of implementation of the second client instead of only atlas but I did that way because it seems safer to avoid any breaking since once it's merged, the other "old" PRs must be rebased to avoid an error that client type is wrong so in that way, if most of the tests of atlas passed it means it works the new implementation, obviously there's a chance that it will fail some testacc like before but it's expected since it hasn't fixed yet

@coderGo93 Thank you for explanation. Feel free to merge, my comment is non blocking.

**Review by thetonymaster — APPROVED** [review]
LGTM just check for the changes you need to do before merging

**Pull request merged**